Reverse words assignment

shrutivanw opened the pull request:
CS Fundamentals assignment to reverse each word in a sentence while preserving white spaces.

shrutivanw committed:
Reverse words assignment

CS Fundamentals assignment to reverse each word in a sentence while preserving white spaces.

kariabancroft commented on README.md:
```diff
@@ -1 +1,9 @@
+* Design and implement a method to reverse each word in a sentence in place. For example, if the method is called with input parameter of "I can be an&nbsp;&nbsp;&nbsp;engineer", the method should update the input string object to have the value "I nac eb na&nbsp;&nbsp;&nbsp;reenigne". Note that the count of white spaces between words is preserved.
+* Share and explain the time and space complexities for your solution.
+* At minimum, your implementation should pass the basic tests.
+**Note**: Do not use Ruby provided functionality for `.reverse` and `.reverse!`. You may use `.length`.
```
I think your desired whitespace between lines 8 & 9 is not preserved -- you probably just need another newline

kariabancroft commented on specs/reverse_words_spec.rb:
```diff
@@ -0,0 +1,82 @@
+
+describe "reverse words" do
+  describe "basic tests" do
+    describe "test 1" do
```
Same comment re: the describe block naming here as the previous code review. I'll assume that you'll incorporate that feedback into the other future assignments and will hold off on reviewing those for now.

shrutivanw committed:
Incorporating Kari's feedback

- a newline added to readme.md so it renders correctly.
- removed inner most nested describe block from basic tests and edge case tests.

shrutivanw commented:
Thank you, @kariabancroft PR updated.

Hamled commented on README.md:
```diff
@@ -1 +1,10 @@
+Remember that a string is an array of characters. Algorithms that worked on restricted arrays will work on strings as well.
+
+## Exercise
+* Design and implement a method to reverse each word in a sentence in place. For example, if the method is called with input parameter of "I can be an&nbsp;&nbsp;&nbsp;engineer", the method should update the input string object to have the value "I nac eb na&nbsp;&nbsp;&nbsp;reenigne". Note that the count of white spaces between words is preserved.
```
A minor quibble: For the input strings we can use code blocks and not have to encode non-breaking space escape sequences: `"I can be an   engineer"`.

Hamled commented on lib/reverse_words.rb:
```diff
@@ -0,0 +1,4 @@
+# A method to reverse each word in a sentence, in place.
+def reverse_words(my_words)
+  my_words << "not implemented"
```
This is a bit confusing to me. It's valid Ruby code but it:
* subtly implies that `my_words` is not a string, since we have only introduced the `<<` operator in the context of arrays (I've actually never seen this used for strings), and
* doesn't adhere to the "in place" requirement, since `<< "not implemented"` appends to the input string.

I would suggest something like:
```ruby
def reverse_words(sentence)
  return sentence
end
```

It doesn't provide the "not implemented" message in the test failure messages, but I think that by the time students are dealing with this they'll have worked with unit tests enough to be able to figure out what's going on when they see `"hello, world" is not equal to ",olleh dlrow"`.

shrutivanw replied:
To share some thinking behind this, and get the benefit of your ideas: 
- In general, all the string manipulation assignments are trying to showcase that behind the scenes, strings are just arrays of characters. 
- Specifically, for this method, I couldn't get the nil and empty test cases to fail without updating the input string (since in the case of nil and empty, the input string should not be altered to reverse the words in place). The "hack" of updating the input string to a random set of characters like "not implemented" made it so the tests failed. If you can think of other, more elegant approach to make the nil and empty test cases to fail, I'm all ears. :-) @Hamled I'll find you sometime this week to chat about this before updating the PR.

shrutivanw replied:
The suggestion you shared for the palindrome exercise, https://github.com/AdaGold/palindrome_check/pull/1#discussion_r160576087 to `raise NotImplementedError` does the trick! Thank you, @Hamled!

Hamled reviewed (changes requested):
Over all this looks good, although I might consider moving the empty string and nil tests into edge cases section, and putting the full sentence test into the basic tests section.

shrutivanw committed:
Improving exercise based on feedback

- The skeleton method to be implemented simply raises NotImplementedError to make all the tests fail.
- nil and empty test cases moved to edge cases

shrutivanw commented:
Thank you @Hamled PR updated to incorporate feedback.
